### PR TITLE
[KAFKA-9826] Handle an unaligned first dirty offset during log cleaning.

### PR DESCRIPTION
**Pull request by steverod**
What
====
In KAFKA-9826, a log whose first dirty offset was past the start of the active segment and past the last cleaned point resulted in an endless cycle of picking the segment to clean and discarding it. Though this didn't interfere with cleaning other log segments, it kept the log cleaner thread continuously busy (potentially wasting CPU and impacting other running threads) and filled the logs with lots of extraneous messages.

This was determined to be because the active segment was getting mistakenly picked for cleaning, and because the `logSegments` code handles (start == end) cases only for (start, end) on a segment boundary: the intent is to return a null list, but if they're not on a segment boundary, the routine returns that segment.

This fix has two parts:
1. It changes logSegments to handle start==end by returning an empty List always.
2. It changes the definition of calculateCleanableBytes to not consider anything past the UncleanableOffset; previously, it would potentially shift the UncleanableOffset to match the firstDirtyOffset even if the firstDirtyOffset was past the firstUncleanableOffset. This has no real effect now in the context of the fix for (1) but it makes the code read more like the model that the code is attempting to follow.

These changes require modifications to a few test cases that handled this particular test case; they were introduced in the context of KAFKA-8764. Those situations are now handled elsewhere in code, but the tests themselves allowed a DirtyOffset in the active segment, and expected an active segment to be selected for cleaning.

An additional unit test for the logSegments call is added. 

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/Log.scala [hunk 1]
```diff
@@ -2132,17 +2132,23 @@ class Log(@volatile private var _dir: File,
...
-      }.getOrElse(segments.headMap(to))
-      view.values.asScala
+    if (from == to) {
+      // Handle non-segment-aligned empty sets
+      List.empty[LogSegment]
+    } else {
+      lock synchronized {
+        val view = Option(segments.floorKey(from)).map { floor =>
+          if (to < floor)
+            throw new IllegalArgumentException(s"Invalid log segment range: requested segments in $topicPartition " +
```
Comment (junrao) [inline core/src/main/scala/kafka/log/Log.scala]: While you are here, it's probably better to consistently throw IllegalArgumentException if `from > to`. Currently, we only throw IllegalArgumentException if we get into this path.

File: core/src/main/scala/kafka/log/Log.scala [hunk 2]
```diff
@@ -2132,17 +2132,23 @@ class Log(@volatile private var _dir: File,
 
   /**
    * Get all segments beginning with the segment that includes "from" and ending with the segment
-   * that includes up to "to-1" or the end of the log (if to > logEndOffset)
+   * that includes up to "to-1" or the end of the log (if to > logEndOffset).
+   *
```
Comment (junrao) [inline core/src/main/scala/kafka/log/Log.scala]: Do we need to add a new line?

**Review by junrao — COMMENTED** [review]
@steverod : Thanks for the patch. Looks good overall. Just a couple of minor comments below.

File: core/src/main/scala/kafka/log/Log.scala [hunk 3]
```diff
@@ -2132,17 +2132,25 @@ class Log(@volatile private var _dir: File,
...
+    if (from == to) {
+      // Handle non-segment-aligned empty sets
+      List.empty[LogSegment]
+    } else if (to < from) {
+      throw new IllegalArgumentException(s"Invalid log segment range: requested segments in $topicPartition " +
+        s"from offset $from which is greater than limit offset $to")
+    } else {
+      lock synchronized {
+        val view = Option(segments.floorKey(from)).map { floor =>
+          if (to < floor)
```
Comment (junrao) [inline core/src/main/scala/kafka/log/Log.scala]: Do we still need this given the test in line 2141?

Comment (steverod) [reply core/src/main/scala/kafka/log/Log.scala:2147]: Removed.

File: core/src/main/scala/kafka/log/Log.scala (hunk 2 above)
Comment (steverod) [reply core/src/main/scala/kafka/log/Log.scala]: Line removed.

File: core/src/main/scala/kafka/log/Log.scala (hunk 1 above)
Comment (steverod) [reply core/src/main/scala/kafka/log/Log.scala]: Done.

File: core/src/main/scala/kafka/log/Log.scala (hunk 3 above)
Comment (junrao) [reply core/src/main/scala/kafka/log/Log.scala]: Was the change to this pushed?

Comment (steverod) [reply core/src/main/scala/kafka/log/Log.scala]: I mistakenly marked this as resolved (then unset it). Change was tested and just pushed, apologies for the unwanted notification!

**Review by junrao — APPROVED** [review]
@steverod : Thanks for the updated PR. LGTM

**Comment by junrao** [comment]
ok to test

**Comment by junrao** [comment]
The failed test seems unrelated. Merging to trunk.

**Pull request merged**